### ISSUE TITLE
Java Grammar: improper method def line and column position for methods with ARRAY_DECLARATOR

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#3145]
Taken from PR #2940

Example 1 with brackets:

```
double[] computeDerivatives()
    |--METHOD_DEF -> METHOD_DEF [85:10] // expected column 4, not 10
    |   |--MODIFIERS -> MODIFIERS [85:10]
    |   |--TYPE -> TYPE [85:10]
    |   |   `--ARRAY_DECLARATOR -> [ [85:10]
    |   |       |--LITERAL_DOUBLE -> double [85:4]
    |   |       `--RBRACK -> ] [85:11]
```

Example 2 with no brackets:

```
double computeDerivatives()
    |--METHOD_DEF -> METHOD_DEF [90:4] // correct line and column
    |   |--MODIFIERS -> MODIFIERS [90:4]
    |   |--TYPE -> TYPE [90:4]
    |   |   `--LITERAL_DOUBLE -> double [90:4]
```

Example 3 with brackets on seperate lines:

```
double
[]
computeDerivatives()
    |--METHOD_DEF -> METHOD_DEF [88:4] // expected line 87, not 88
    |   |--MODIFIERS -> MODIFIERS [88:4]
    |   |--TYPE -> TYPE [88:4]
    |   |   `--ARRAY_DECLARATOR -> [ [88:4]
    |   |       |--LITERAL_DOUBLE -> double [87:4]
    |   |       `--RBRACK -> ] [88:5]
```

I expected `METHOD_DEF` to signify the exact line and column position of the start of the method definition. In this case, example 1's column position 10 is wrong and it should be 4, which we see for when there is no brackets in example 2. Even if we go by the line number and find the start of the line, the 3rd example shows it can be wrong too.

Full Example:

```
$ cat TestClass.java
public class TestClass {
    double[] computeDerivatives() { return null; }
}

$ java -jar checkstyle-6.17-all.jar TestClass.java -T
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|  `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> TestClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:23]
    |--LCURLY -> { [1:23]
    |--METHOD_DEF -> METHOD_DEF [2:10] // expected 2:4 
    |  |--MODIFIERS -> MODIFIERS [2:10]
    |  |--TYPE -> TYPE [2:10]
    |  |  `--ARRAY_DECLARATOR -> [ [2:10]
    |  |      |--LITERAL_DOUBLE -> double [2:4] // as seen here
    |  |      `--RBRACK -> ] [2:11]
    |  |--IDENT -> computeDerivatives [2:13]
    |  |--LPAREN -> ( [2:31]
    |  |--PARAMETERS -> PARAMETERS [2:32]
    |  |--RPAREN -> ) [2:32]
    |  `--SLIST -> { [2:34]
    |      |--LITERAL_RETURN -> return [2:36]
    |      |  |--EXPR -> EXPR [2:43]
    |      |  |  `--LITERAL_NULL -> null [2:43]
    |      |  `--SEMI -> ; [2:47]
    |      `--RCURLY -> } [2:49]
    `--RCURLY -> } [3:0]
```
-------

**Migration notes**: please read at https://github.com/checkstyle/checkstyle/issues/3145#issuecomment-850996021

**Comment by harsh-kukreja** [comment]
Hey @romani / @rnveach how do i go about changing the grammar? Can you refer me to some documentations?

**Comment by rnveach** [comment]
There is no documentation in checkstyle if that is what you are referring to. We use ANTLR 2.7 so you can google online and learn about ANTLR and find its documentation.

**Comment by harsh-kukreja** [comment]
Okay @rnveach thanks for the reply

**Comment by nmancus1** [comment]
I am on this.

**Comment by nmancus1** [comment]
How about this:

```bash

➜  src cat MyClass.java
public class MyClass {
    double[] computeDerivatives() {
        return null;
    }

    double[][] computeIntegration() {
        return null;
    }
}

➜  src java -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.42-SNAPSHOT-all.jar -t MyClass.java

CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> MyClass [1:13]
`--OBJBLOCK -> OBJBLOCK [1:21]
    |--LCURLY -> { [1:21]
    |--METHOD_DEF -> METHOD_DEF [2:4]
    |   |--MODIFIERS -> MODIFIERS [2:4]
    |   |--TYPE -> TYPE [2:4]
    |   |   |--LITERAL_DOUBLE -> double [2:4]
    |   |   |--ARRAY_DECLARATOR -> [ [2:10]
    |   |   `--RBRACK -> ] [2:11]
    |   |--IDENT -> computeDerivatives [2:13]
    |   |--LPAREN -> ( [2:31]
    |   |--PARAMETERS -> PARAMETERS [2:32]
    |   |--RPAREN -> ) [2:32]
    |   `--SLIST -> { [2:34]
    |       |--LITERAL_RETURN -> return [3:8]
    |       |   |--EXPR -> EXPR [3:15]
    |       |   |   `--LITERAL_NULL -> null [3:15]
    |       |   `--SEMI -> ; [3:19]
    |       `--RCURLY -> } [4:4]
    |--METHOD_DEF -> METHOD_DEF [6:4]
    |   |--MODIFIERS -> MODIFIERS [6:4]
    |   |--TYPE -> TYPE [6:4]
    |   |   |--LITERAL_DOUBLE -> double [6:4]
    |   |   |--ARRAY_DECLARATOR -> [ [6:10]
    |   |   |--RBRACK -> ] [6:11]
    |   |   |--ARRAY_DECLARATOR -> [ [6:12]
    |   |   `--RBRACK -> ] [6:13]
    |   |--IDENT -> computeIntegration [6:15]
    |   |--LPAREN -> ( [6:33]
    |   |--PARAMETERS -> PARAMETERS [6:34]
    |   |--RPAREN -> ) [6:34]
    |   `--SLIST -> { [6:36]
    |       |--LITERAL_RETURN -> return [7:8]
    |       |   |--EXPR -> EXPR [7:15]
    |       |   |   `--LITERAL_NULL -> null [7:15]
    |       |   `--SEMI -> ; [7:19]
    |       `--RCURLY -> } [8:4]
    `--RCURLY -> } [9:0]



```

**Comment by rnveach** [comment]
> The weird line/ column numbers are from our classTypeSpec and builtinTypeSpec productions

It is actually caused because we are creating nodes not built from the source. There is no text that signifies the start of `METHOD_DEF`.
Instead, when the ANTLR gives us no exact line/column, we try to calculate it's position by looking at it's children and siblings.
See https://github.com/checkstyle/checkstyle/blob/13481f2c410e4944ecf5ab93ec49948a523a0c82/src/main/java/com/puppycrawl/tools/checkstyle/DetailAstImpl.java#L220

If we can resolve the issue in ANTLR, lets go for it.

>    |   |--TYPE -> TYPE [2:4]
>    |   |   |--LITERAL_DOUBLE -> double [2:4]
>    |   |   |--ARRAY_DECLARATOR -> [ [2:10]
>    |   |   `--RBRACK -> ] [2:11]

I would assume `RBRACK` should be a child of `ARRAY_DECLARATOR`, otherwise the change makes sense to me.

**Comment by nmancus1** [comment]
> It is actually caused because we are creating nodes not built from the source. 

> because we create the imaginary/modified node, e.g. `= #(#[METHOD_DEF,"METHOD_DEF"],`

That is correct, as I stated above, but the strange production rules for `classTypeSpec` and `builtinTypeSpec` is why we do not use the node that comes first syntactically; both of those rules make the `LBRACK` the root node of that production, so it is that token's location that is being used in this case.

> I would assume `RBRACK` should be a child of `ARRAY_DECLARATOR`

Yes, that is a good idea I agree.

**Comment by nmancus1** [comment]
This fix will change the structure of the `TYPE` ast, allowing us to get the true first column of the `METHOD_DEF` and `VARIABLE_DEF` ast.

### Example:

```bash

➜  src cat Test.java
public class Test {
    int[] method() {
        return new int[3];
    }
}


```

AST before this fix:

```bash

➜  src java -jar checkstyle-8.42-all.jar -t Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--METHOD_DEF -> METHOD_DEF [2:7] <- AST really starts at column 4, not 7
    |   |--MODIFIERS -> MODIFIERS [2:7]
    |   |--TYPE -> TYPE [2:7]
    |   |   `--ARRAY_DECLARATOR -> [ [2:7]  <- "first" element in METHOD_DEF gives column position
    |   |       |--LITERAL_INT -> int [2:4]            <- method return type is nested in '[' and ']'
    |   |       `--RBRACK -> ] [2:8]
    |   |--IDENT -> method [2:10]
    |   |--LPAREN -> ( [2:16]
    |   |--PARAMETERS -> PARAMETERS [2:17]
    |   |--RPAREN -> ) [2:17]
    |   `--SLIST -> { [2:19]
    |       |--LITERAL_RETURN -> return [3:8]
    |       |   |--EXPR -> EXPR [3:15]
    |       |   |   `--LITERAL_NEW -> new [3:15]
    |       |   |       |--LITERAL_INT -> int [3:19]
    |       |   |       `--ARRAY_DECLARATOR -> [ [3:22]
    |       |   |           |--EXPR -> EXPR [3:23]
    |       |   |           |   `--NUM_INT -> 3 [3:23]
    |       |   |           `--RBRACK -> ] [3:24]
    |       |   `--SEMI -> ; [3:25]
    |       `--RCURLY -> } [4:4]
    `--RCURLY -> } [5:0]


```

AST after this fix:

```bash

➜  src java -jar ~/IdeaProjects/checkstyle/target/checkstyle-8.43-SNAPSHOT-all.jar -t Test.java
CLASS_DEF -> CLASS_DEF [1:0]
|--MODIFIERS -> MODIFIERS [1:0]
|   `--LITERAL_PUBLIC -> public [1:0]
|--LITERAL_CLASS -> class [1:7]
|--IDENT -> Test [1:13]
`--OBJBLOCK -> OBJBLOCK [1:18]
    |--LCURLY -> { [1:18]
    |--METHOD_DEF -> METHOD_DEF [2:4] <- now AST shows correct column
    |   |--MODIFIERS -> MODIFIERS [2:4]
    |   |--TYPE -> TYPE [2:4]
    |   |   |--LITERAL_INT -> int [2:4]  <- true first element in METHOD_DEF gives column position, no longer nested
    |   |   `--ARRAY_DECLARATOR -> [ [2:7] 
    |   |       `--RBRACK -> ] [2:8]
    |   |--IDENT -> method [2:10]
    |   |--LPAREN -> ( [2:16]
    |   |--PARAMETERS -> PARAMETERS [2:17]
    |   |--RPAREN -> ) [2:17]
    |   `--SLIST -> { [2:19]
    |       |--LITERAL_RETURN -> return [3:8]
    |       |   |--EXPR -> EXPR [3:15]
    |       |   |   `--LITERAL_NEW -> new [3:15]
    |       |   |       |--LITERAL_INT -> int [3:19]
    |       |   |       `--ARRAY_DECLARATOR -> [ [3:22]
    |       |   |           |--EXPR -> EXPR [3:23]
    |       |   |           |   `--NUM_INT -> 3 [3:23]
    |       |   |           `--RBRACK -> ] [3:24]
    |       |   `--SEMI -> ; [3:25]
    |       `--RCURLY -> } [4:4]
    `--RCURLY -> } [5:0]

```

So, custom checks may need to make changes like: https://github.com/sevntu-checkstyle/sevntu.checkstyle/pull/845/files#diff-cf22a9a4229906d2b13b60e76c1195bc20e48f81235910bf3764cb258e7be28f or https://github.com/checkstyle/checkstyle/pull/9922/files#diff-a2706cddb4adc2186effbc0df613b9be43b717a3c3bc5d77b59145a2834e5bcd . 

In summary, instead of expecting the return type of a variable or method AST to be nested in the `ARRAY_DECLARATOR` brackets, it is now the first child of the `TYPE` AST.

**Comment by romani** [comment]
fix is merged, huge thanks to @nmancus1  for work on this.
